autogen/oai/cohere.py

```
@@ -0,0 +1,427 @@
+        # search_queries_only - to control whether only search queries are used, we're not using connectors so ignoring.
+        # documents - a list of documents that can be used to support the chat. Perhaps useful in the future for RAG.
+        # citation_quality - used for RAG flows and dependent on other parameters we're ignoring.
+        # max_input_tokens - limits input tokens, not needed.
+        # stop_sequences - used to stop generation, not needed.
+
+        return cohere_params
+
+    def create(self, params: Dict) -> ChatCompletion:
+
+        messages = params.get("messages", [])
+
+        # Parse parameters to the Cohere API's parameters
+        cohere_params = self.parse_params(params)
+
+        # Convert AutoGen messages to Cohere messages
+        cohere_messages, preamble, final_message = oai_messages_to_cohere_messages(messages, params, cohere_params)
+
+        cohere_params["chat_history"] = cohere_messages
+        cohere_params["message"] = final_message
+        cohere_params["preamble"] = preamble
+
+        # We use chat model by default
+        client = Cohere(api_key=self.api_key)
+
+        # Token counts will be returned
+        prompt_tokens = 0
+        completion_tokens = 0
+        total_tokens = 0
+
+        # Streaming tool call recommendations
+        # streaming_tool_calls = []
+
+        # Stream if in parameters
+        streaming = True if "stream" in params and params["stream"] else False
+        cohere_finish = ""
+
+        max_retries = 5
+        for attempt in range(max_retries):
+            ans = None
+            try:
+                if streaming:
+                    response = client.chat_stream(**cohere_params)
+                else:
+                    response = client.chat(**cohere_params)
+            except Exception as e:
+                raise RuntimeError(f"Cohere exception occurred: {e}")
```

i dont think this is required. would love some feedback from @yiranwu0 @qingyun-wu

---

@Hk669, which part you think is not required?

---

the `max_retries` loop, i dont think we follow the same for the other clients. let it throw an exception. 

cc @marklysze

---

Gemini has a [retries loop](https://github.com/microsoft/autogen/blob/8133b7de22a7b9a7c84cee31bbd963798efef753/autogen/oai/gemini.py#L193) but not all do. If we want to standardise on no loop, that's okay with me.

---

for now, let's maintain the current approach and adhere to the client's SDK as specified. thank you @marklysze